Fix an issue with Classic Text fields in a repeater

kienstra opened the pull request:
As Rob [pointed out](https://github.com/getblocklab/block-lab/issues/465#issue-515808255), only the 1st Classic Text field in a repeater works.

This looks to be because all Classic Text controls in the same Repeater were using the same `editorId`. So this creates an `editorId` that's aware of the `rowIndex`.

### Steps to reproduce

1. In the Block Lab 'Edit Block' UI, add a Repeater control to a block, and add a Classic Text control to that
2. Create a new post, and add that block
3. The Classic Text control should work fine
4. Add another row to the Repeater
5. Expected: The Classic Text control in that 2nd row should work
6: Actual: It's not possible to edit that Classic Text control

Fixes #465

kienstra committed:
Fix an issue with Classic Text fields in a repeater

As Rob pointed out, only the 1st Classic Text
field in a repeater works.
This looks to be because they were using the
same editorId.
So create a different editorId, using the rowIndex,
if that exists.

kienstra commented:
Also, I tested that multiple Classic Text controls in the same block (not in a repeater) also work. They worked before and after this PR.

kienstra committed:
Account for a rowIndex of 0

Before, this wouldn't have included the rowIndex in the editorId
if it was 0.
Props @oscarssanchez.

kienstra commented:
**Request For Review**

Hi @RobStino and @lukecarbis,
Could you please review this? Hope you had a great weekend.

Thanks!

jcrooke commented:
I can confirm this is now fixed 👍

kienstra commented:
Nice, thanks for testing this!

lukecarbis commented:
Code looks good to me, but I'm still having issues (slightly different that the issues I had before these changes).

https://fromsmash.com/fXP5zftmWr-c0

kienstra commented:
Hi @lukecarbis,
OK, thanks for bringing that up. I'm a little sick today, so I probably won't have time to work on this today.

kienstra commented:
Sorry I didn't get to this today. I'll work on it tomorrow if I'm not feeling worse, or Sunday (in the US) at the latest.

kienstra commented:
Thanks for the good video of this. I could also reproduce this, but haven't figured out how to fix it.

I'll work on it more tomorrow (Sunday in the US).

kienstra committed:
Add a key to the containing <div> (though this doesn't fix the bug)

This looks to address a console error
from a key not being present.

kienstra commented:
I'm working on this now. It's a tough one 😄

kienstra committed:
Address the issue Luke found where Classic Text fields were removed

It looks like onChange()
was using outdated parent block attributes.
So this instead uses select( 'core/block-editor' )
to get the latest attributes.

kienstra committed:
Add support in the Classic Text component for the default field

Strangely, I didn't do this before.
Though the default appeared on the front-end
and in the <ServerSideRender>.
It just didn't appear in the TinyMCE before.

kienstra commented:
I think the [commit above](https://github.com/getblocklab/block-lab/pull/466/commits/14532fcd6c3337796bc1023e3dd320573469c90b) fixes the [issue with disappearing Classic Text fields](https://github.com/getblocklab/block-lab/pull/466#issuecomment-550134215). But I need to test this a lot, as it has edits to `onChange()`, which every control uses to save its values.

kienstra commented on js/blocks/components/fields.js:
```diff
@@ -59,33 +60,31 @@ const Fields = ( { fields, parentBlockProps, parentBlock, rowIndex } ) => {
 		 * This function is passed to the control so that the control can save the value,
 		 * depending on whether the control is in a repeater row or not.
+		 * This gets the latest parentAttributes by using select( 'core/block-editor' ),
+		 * as the parentBlockProps.attributes in the outer scope may be outdated.
```
It looks like the issue with the disappearing Classic Text fields was that it was using outdated `parentBlockProps.attributes`. So this gets them from the store.

kienstra commented on js/blocks/components/fields.js:
```diff
@@ -59,33 +60,31 @@ const Fields = ( { fields, parentBlockProps, parentBlock, rowIndex } ) => {
-				 * Copy the rows array, so the change is recognized.
-				 * @see https://github.com/WordPress/gutenberg/issues/7016#issuecomment-396094836
-				 */
-				const rowsCopy = rows.slice();
```
It looks like this `rows.slice()` hack isn't needed anymore to allow saving repeater values. But more testing would be good.

kienstra replied:
I tested saving all of the controls in a repeater, and didn't see an issue.

lukecarbis replied:
I also tested, everything seemed to work.

kienstra commented on js/blocks/components/fields.js:
```diff
@@ -118,9 +117,8 @@ const Fields = ( { fields, parentBlockProps, parentBlock, rowIndex } ) => {
 
 		return !! Control && (
-			<div className={ getClassName( field ) }>
+			<div className={ getClassName( field ) } key={ `${ field.name }-control-${ rowIndex }` }>
```
This isn't related to the bug, but it prevents a console warning:

```
react-dom.js:539 Warning: Each child in a list should have a unique "key" prop. See https://fb.me/react-warning-keys for more information.
    in Fields (created by FormControls)
    in FormControls (created by Edit)
    in div (created by Edit)
    in div (created by Edit)
    in Edit (created by edit)
```

RobStino commented:
Hey @kienstra 
Just been testing this and the repeater work looks good. However classic text outside of the repeater is having an issue. :(
Testing included:
- new block
- Repeater field with a classic text subfield and a normal text subfield
- A normal non-repeater classic field and a number field

Video below of what I see:
https://d.pr/v/Hmwtnb

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Sorry for the delay. Could you please review this when you have a chance?

In my testing, saving all of the control types still works, both with and without a repeater.

Of course, before the upcoming release, we should do at least as much regression testing as usual.

Thanks, Luke!

kienstra committed:
Avoid a collision in the editorId when there's a repeater and non-repeater

Before, it was possible to have the same
editorId if there is a field like
classic-2,
and there's a repeater with 3 rows of Classic Text fields.
In that case, the 3rd row would have an editorId
of bl-classic-text-classic-2,
just like the field.

kienstra commented:
Hi @RobStino,
Thanks for testing this. Does the latest commit fix the issue for you?

I also saw the issue locall after watching your screencast, and it looks like that commit takes care of it.

RobStino commented:
Hey @kienstra 
Thanks for that fix. Working correctly now.

lukecarbis commented:
Looks good to me too. There's one other bug that I noticed: Re-ordering the classic fields inside a repeater doesn't work.

I'd be happy to open a new issue for that, if you wanted to merge this one, though.

kienstra commented:
>I'd be happy to open a new issue for that, if you wanted to merge this one, though.

Yeah, that'd be great if you could open another issue for that.
